### PR TITLE
refactor(gateway): support sushiswap on polygon, mumbai, goerli

### PR DESCRIPTION
**Pull request by gzliudan**
**Before submitting this PR, please make sure**:

- [X] Your code builds clean without any errors or warnings
- [X] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Support sushiswap on polygon and mumbai chain

**Tests performed by the developer**:

Query network status of polygon

```shell
curl -s -X GET -k --key $GATEWAY_KEY --cert $GATEWAY_CERT "${SERVER}/network/status?chain=polygon&network=mainnet" | jq
```

---

Query network status of mumbai

```shell
curl -s -X GET -k --key $GATEWAY_KEY --cert $GATEWAY_CERT "${SERVER}/network/status?chain=polygon&network=mumbai" | jq
```

---

Query token list of mumbai

```shell
curl -s -X GET -k --key $GATEWAY_KEY --cert $GATEWAY_CERT "${SERVER}/network/tokens?chain=polygon&network=mumbai" | jq
```

---

Query token list of polygon

```shell
curl -s -X GET -k --key $GATEWAY_KEY --cert $GATEWAY_CERT "${SERVER}/network/tokens?chain=polygon&network=mainnet" | jq
```

---

query sushiswap mumbai buy price

```shell
curl -s -X POST -k --key $GATEWAY_KEY --cert $GATEWAY_CERT -H "${HEADER}" "${SERVER}/amm/price" -d '{
    "connector": "sushiswap",
    "chain": "polygon",
    "network": "mumbai",
    "base": "WBTC2",
    "quote": "USDC2",
    "amount": "1",
    "side": "BUY"
}' | jq
```

---

query sushiswap mumbai sell price

```shell
curl -s -X POST -k --key $GATEWAY_KEY --cert $GATEWAY_CERT -H "${HEADER}" "${SERVER}/amm/price" -d '{
    "connector": "sushiswap",
    "chain": "polygon",
    "network": "mumbai",
    "base": "WBTC2",
    "quote": "USDC2",
    "amount": "1",
    "side": "SELL"
}' | jq
```

---

trade of mumbai sushiswap buy

```shell
curl -s -X POST -k --key $GATEWAY_KEY --cert $GATEWAY_CERT -H "${HEADER}" "${SERVER}/amm/trade" -d '{
    "address": "'"$ETH_ADDRESS"'",
    "base": "WBTC2",
    "quote": "USDC2",
    "amount": "1",
    "side": "BUY",
    "chain": "polygon",
    "network": "mumbai",
    "connector": "sushiswap"
}' | jq
```

---

trade of mumbai sushiswap sell

```shell
curl -s -X POST -k --key $GATEWAY_KEY --cert $GATEWAY_CERT -H "${HEADER}" "${SERVER}/amm/trade" -d '{
    "address": "'"$ETH_ADDRESS"'",
    "base": "WBTC2",
    "quote": "USDC2",
    "amount": "1",
    "side": "SELL",
    "chain": "polygon",
    "network": "mumbai",
    "connector": "sushiswap"
}' | jq
```

---

gateway connect sushiswap

**Tips for QA testing**:

The adderss of SushiSwapRouter on [polygon](https://polygonscan.com/address/0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506#writeContract) and [mumbai](https://mumbai.polygonscan.com/address/0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506#writeContract)  is `0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506`.

My test tokens:
| Symbol  | Decimals | Address                                                                                                                                         | 
| :---- | ---: | :------------------------------------------------------------------------------------------------------------------------------------------- |
| WBTC2 |    8 | [0x04597a1486b4F9d8284A4771703C1808D2a29b81](https://mumbai.polygonscan.com/address/0x04597a1486b4F9d8284A4771703C1808D2a29b81#readContract) |
| YFI2  |   18 | [0xa79391e7B2b48a4Eb7F2228ef625a32716eE4E93](https://mumbai.polygonscan.com/address/0xa79391e7B2b48a4Eb7F2228ef625a32716eE4E93#readContract) |
| MKR2  |   18 | [0xc16d215c621bbfd12999c9f456dab44db755ddd5](https://mumbai.polygonscan.com/address/0xc16d215c621bbfd12999c9f456dab44db755ddd5#readContract) |
| AAVE2 |   18 | [0x34006a403f5aba9f247a284e6278fbcb7c00df38](https://mumbai.polygonscan.com/address/0x34006a403f5aba9f247a284e6278fbcb7c00df38#readContract) |
| UNI2  |   18 | [0x0a696e15754f9de196a939fecb8c412812d5881d](https://mumbai.polygonscan.com/address/0x0a696e15754f9de196a939fecb8c412812d5881d#readContract) |
| USDC2 |    6 | [0xb6ae5ac95088f80f43a52f6be2b44906492b0d98](https://mumbai.polygonscan.com/address/0xb6ae5ac95088f80f43a52f6be2b44906492b0d98#readContract) |

Anyone can mint tokens, no quantity limit.

### CODE REVIEW

**Comment by fengtality** [comment]
@gzliudan Sorry for the late reply. We've made some architectural changes to Gateway recently and added BNB Chain support for SushiSwap. Do you want to update this PR for the latest `development` code?

**Comment by gzliudan** [comment]
Ok, I can update this PR next week.

**Comment by gzliudan** [comment]
This PR is updated on the latest development branch today.

File: gateway/src/chains/polygon/polygon_tokens_mumbai.json [hunk 1]
```diff
@@ -4,20 +4,56 @@
     {
       "symbol": "WMATIC",
       "chainId": 80001,
-      "address": "0x9c3c9283d3e44854697cd22d3faa240cfb032889",
```
Comment (fengtality) [inline gateway/src/chains/polygon/polygon_tokens_mumbai.json:7]: Please remove the changes to addresses unless changes are really needed. This prevents potential security issues -thanks

File: gateway/src/chains/polygon/polygon_tokens_mumbai.json [hunk 2]
```diff
@@ -4,20 +4,56 @@
...
+      "address": "0x9c3C9283D3e44854697Cd22D3Faa240Cfb032889",
       "decimals": 18
     },
     {
       "symbol": "WETH",
       "chainId": 80001,
-      "address": "0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa",
+      "address": "0xA6FA4fB5f76172d178d61B04b0ecd319C5d1C0aa",
```
Comment (fengtality) [inline gateway/src/chains/polygon/polygon_tokens_mumbai.json:13]: same as above

File: gateway/test/chains/polygon/quickswap/quickswap.test.ts [hunk 3]
```diff
@@ -21,13 +21,13 @@ let quickswap: Quickswap;
 
 const WMATIC = new Token(
   80001,
-  '0x9c3c9283d3e44854697cd22d3faa240cfb032889',
```
Comment (fengtality) [inline gateway/test/chains/polygon/quickswap/quickswap.test.ts:24]: same as above

File: gateway/src/templates/sushiswap.yml [hunk 4]
```diff
@@ -19,8 +19,16 @@ contractAddresses:
...
+      sushiswapRouterAddress: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'
+    mumbai:
+      sushiswapRouterAddress: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'
   binance-smart-chain:
     mainnet:
       sushiswapRouterAddress: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'
     testnet:
-      sushiswapRouterAddress: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'
\ No newline at end of file
+      sushiswapRouterAddress: '0x1b02dA8Cb0d097eB8D57A175b88c7D8b47997506'
```
Comment (fengtality) [inline gateway/src/templates/sushiswap.yml:33]: same as above

File: gateway/test/chains/polygon/quickswap/quickswap.test.ts [hunk 5]
```diff
@@ -21,13 +21,13 @@ let quickswap: Quickswap;
...
+  '0x9c3C9283D3e44854697Cd22D3Faa240Cfb032889',
   18,
   'WMATIC'
 );
 const WETH = new Token(
   80001,
-  '0xa6fa4fb5f76172d178d61b04b0ecd319c5d1c0aa',
+  '0xA6FA4fB5f76172d178d61B04b0ecd319C5d1C0aa',
```
Comment (fengtality) [inline gateway/test/chains/polygon/quickswap/quickswap.test.ts:30]: same as above

**Review by fengtality — CHANGES_REQUESTED** [review]

**Comment by gzliudan** [comment]
@fengtality @rapcmia OK, the files gateway/test/chains/polygon/quickswap/quickswap.test.ts and gateway/src/chains/polygon/polygon_tokens_mumbai.json are retained unchanged now.

**Comment by rapcmia** [comment]
PR update:
- Setup source client and gateway successfully
- Tested on mainnet and connected wallet successfully
- Balance display correctly and connector-tokens parameter works as expected 
- Setup simple amm arb
   - Setup config with sushiswap-polygon mainnet and paper_trade connector 
   - Approval success and markets are ready displayed
   - Profitability for no arbitrage displays correctly
   - Difference between the prices of `status` and exchange swap price is close enough
- Build docker image for client and gateway for local build tests
   - Tested on mumbai testnet using WMATIC-DAI
   - Same results achieved as source build 👍 
- Test endpoint:
   - port, wallet and balance check: ok 
   - approval, prices and trade: ok 

Note: _There are noticeable issues on polygon chain on tx execution and not related to this PR_

Pending:
Currently running long term for test for 3-5hrs

**Comment by rapcmia** [comment]
PR good to be approved however would need pass the GitHub CI:
image

**Comment by rapcmia** [comment]
Hi @gzliudan good day to you
We have not received any response from the previous comment, just checking if you can provide us a timeframe of a possible new commit? PR is good to be approve but getting failed GitHub Ci due to not enough coverages:
image

**Comment by gzliudan** [comment]
@rapcmia I have no experence on TypeScript, so I can modify some codes, but have no ability to write test codes.

**Comment by rapcmia** [comment]
Hi @gzliudan good day to you
This PR has been selected as one of top 5 to be added to the next release `1.13.0` from this [poll](https://snapshot.org/#/hbot.eth/proposal/0x5e023da8e21855ec4bb34fcbc124513831b93668a10f788a7d965e71a7b67d28). Please also be advised that we have changes on gateway which is now moved to a new [repository](https://github.com/hummingbot/gateway). At the moment the PR needs to be updated base on the recent changes, current does not display any for polygon / mumbai support for sushiswap

image

**Comment by gzliudan** [comment]
I will submit a new PR to gateway [repository](https://github.com/hummingbot/gateway) next week.

> Hi @gzliudan good day to you This PR has been selected as one of top 5 to be added to the next release `1.13.0` from this [poll](https://snapshot.org/#/hbot.eth/proposal/0x5e023da8e21855ec4bb34fcbc124513831b93668a10f788a7d965e71a7b67d28). Please also be advised that we have changes on gateway which is now moved to a new [repository](https://github.com/hummingbot/gateway). At the moment the PR needs to be updated base on the recent changes, current does not display any for polygon / mumbai support for sushiswap
> 
> image

**Comment by gzliudan** [comment]
@rapcmia  Done: https://github.com/hummingbot/gateway/pull/16

**Pull request closed without merging**